xdsl/interactive/app.py

```diff
@@ -0,0 +1,256 @@
+from collections.abc import Callable
+from io import StringIO
+
+# pyright: ignore[reportMissingTypeStubs, reportGeneralTypeIssues]
+from pyclip import copy as pyclip_copy
+from textual import events, on
+from textual.app import App, ComposeResult
+from textual.containers import Horizontal, ScrollableContainer, Vertical
+from textual.widgets import Button, Label, SelectionList, TextArea
+
+from xdsl.dialects import builtin
+from xdsl.ir import MLContext
+from xdsl.parser import Parser
+from xdsl.passes import ModulePass, PipelinePass
+from xdsl.printer import Printer
+from xdsl.tools.command_line_tool import get_all_dialects, get_all_passes
+
+"""
+This app allows you to paste xDSL IR into the Input TextArea, select a pass (or multiple passes) to be applied on the input IR
+and subsequently the IR generated by the application of the selected pass(es) is displayed in the Output TextArea. The selected
+passes are displayed in the top right "Selecred passes/query" box. The "Condense" button filters the pass selection list to
+display passes that change the IR, or require arguments to be executed (and thus "may" change the IR). The "Uncondense" button
+returns the selection list to contain all the passes. The "Clear Passes" Button removes the application of the selected passes.
+The "Copy Query" Button allows the user to copy the selected passes/query that they have so far selected (i.e. copy the top right
+box).
+
+This app is still under construction.
+"""
+pyclip_copy: Callable[[str], None] = pyclip_copy
+
+
+# Function that takes the input IR, the list of passes to be applied, and applies the passes to the IR.
+# Returns the module (after pass is applied).
+def transform_input(
+    input_text: str, passes: list[type[ModulePass]]
+) -> builtin.ModuleOp:
+    ctx = MLContext(True)
+    for dialect in get_all_dialects():
+        ctx.load_dialect(dialect)
+
+    parser = Parser(ctx, input_text)
+    module = parser.parse_module()
+
+    pipeline = PipelinePass([p() for p in passes])
+    pipeline.apply(ctx, module)
+
+    return module
+
+
+# Used to check if a pass has had an effect on the IR (i.e. has changed the IR)
+def check_if_pass_changes_IR(
+    input: builtin.ModuleOp, ctx: MLContext, one_pass: ModulePass
+) -> bool:
+    cloned_module = input.clone()
+    cloned_ctx = ctx.clone()
+    one_pass.apply(cloned_ctx, cloned_module)
+    return not input.is_structurally_equivalent(cloned_module)
+
+
+# Used to prevent users from being able to change/alter the Output TextArea
+class OutputTextArea(TextArea):
+    async def _on_key(self, event: events.Key) -> None:
+        event.prevent_default()
+
+
+# Class buildling the Experimental Interactive Compilation App.
+class InputApp(App[None]):
```

Comment:
Textual encourages state to be held in reactive variables on the App and Widget classes, and the UI to be derived from those values as they change. Could you please transition to that system?

Reply:
This should also make it easier to evolve the app over time